taichi/transforms/binary_op_simplify.cpp

```diff
@@ -0,0 +1,105 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/program/program.h"
+
+TLANG_NAMESPACE_BEGIN
+
+class BinaryOpSimp : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+  bool fast_math;
+  DelayedIRModifier modifier;
+
+  explicit BinaryOpSimp(bool fast_math_)
+      : BasicStmtVisitor(), fast_math(fast_math_) {
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    // swap lhs and rhs if lhs is a const and op is commutative
+    auto const_lhs = stmt->lhs->cast<ConstStmt>();
+    if (const_lhs && is_commutative(stmt->op_type) &&
+        !stmt->rhs->is<ConstStmt>()) {
+      auto rhs_stmt = stmt->rhs;
+      stmt->lhs = rhs_stmt;
+      stmt->rhs = const_lhs;
+    }
+    if (!fast_math) {
+      return;
+    }
+    auto binary_lhs = stmt->lhs->cast<BinaryOpStmt>();
+    auto const_rhs = stmt->rhs->cast<ConstStmt>();
+    if (!binary_lhs || !const_rhs) {
+      return;
+    }
+    auto const_lhs_rhs = binary_lhs->rhs->cast<ConstStmt>();
+    if (!const_lhs_rhs) {
+      return;
```

Comment:
There's a bug here, it should also check that the `stmt->lhs->lhs` is not a const. Otherwise, we are swapping between consts, causing loops and therefore CI failed.